libsql/internal/http/hranaV2/hranaV2.go

```diff
@@ -286,7 +287,119 @@ func sendPipelineRequest(ctx context.Context, msg *hrana.PipelineRequest, url st
 	return result, false, nil
 }
 
+func (h *hranaV2Conn) executeMsg(ctx context.Context, query string, msg *hrana.PipelineRequest) (*hrana.PipelineResponse, error) {
+	result, err := h.sendPipelineRequest(ctx, msg, false)
+	if err != nil {
+		return nil, fmt.Errorf("failed to execute SQL: %s\n%w", query, err)
+	}
+
+	for _, r := range result.Results {
+		if r.Error != nil {
+			return nil, fmt.Errorf("failed to execute SQL: %s\n%s", query, r.Error.Message)
+		}
+		if r.Response == nil {
+			return nil, fmt.Errorf("failed to execute SQL: %s\n%s", query, "no response received")
+		}
+	}
+	return result, nil
+}
+
+func (h *hranaV2Conn) executeInChunks(ctx context.Context, query string, wantRows bool) (*hrana.PipelineResponse, error) {
```

Comment:
To be honest, I find current implementation pretty complicated...
My suggestion is the following with following direction for improvements:
1. We can introduce `chunker` which will group statements and we will iterate over it's result rather result or `iterator`
2. I suggest to treat all iterations in the same main loop without explicit extraction of the first and last step (as we have now)
```go
type sqlChunker struct {
	chunk    []string
	iterator *sqliteparserutils.StatementIterator
}

func (c *sqlChunker) Next() (chunk []string, isEOF bool) {
	c.chunk = c.chunk[:0]
	var stmt string
	for {
		stmt, _, isEOF = c.iterator.Next()
		if isEOF {
			break
		}
		c.chunk = append(c.chunk, stmt)
	}
	return c.chunk, isEOF
}

func (h *hranaV2Conn) executeInChunks(ctx context.Context, query string, wantRows bool) (*hrana.PipelineResponse, error) {
	iterator := sqliteparserutils.CreateStatementIterator(query)
	chunker := sqlChunker{iterator: iterator, chunk: make([]string, 0, 4096)}

	sqlBatch, isEOF := chunker.Next()
	// todo (sivukhin, 2024-07-22): what to do if query has no non-empty statements?
	if isEOF && len(sqlBatch) == 1 {
		executeStreamRequest, err := hrana.ExecuteStream(sqlBatch[0], nil, wantRows)
		if err != nil {
			return nil, fmt.Errorf("failed to create ExecuteStreamRequest: %w", err)
		}
		return h.executeMsg(ctx, query, &hrana.PipelineRequest{Requests: []hrana.StreamRequest{*executeStreamRequest}})
	}

	hranaBatch := &hrana.Batch{Steps: make([]hrana.BatchStep, 0, len(sqlBatch)+2)}
	result := &hrana.PipelineResponse{}

	for chunkIdx := 0; ; chunkIdx++ {
		hranaBatch.Steps = hranaBatch.Steps[:0]
		if chunkIdx == 0 {
			hranaBatch.Add(hrana.NewStmt("BEGIN", false), nil)
		}
		itemsStartIdx := len(sqlBatch)
		for _, sqlItem := range sqlBatch {
			hranaBatch.Add(hrana.NewStmt(sqlItem, wantRows), nil)
		}
		itemsEndIdx := len(sqlBatch)
		if isEOF {
			hranaBatch.Add(hrana.NewStmt("COMMIT", false), nil)
		}

		request := &hrana.PipelineRequest{}
		request.Add(hrana.StreamRequest{Type: "batch", Batch: hranaBatch})
		response, err := h.executeMsg(ctx, query, request)
		if err != nil {
			h.closeStream()
			return nil, fmt.Errorf("failed to execute batch %d: %w", chunkIdx, err)
		}
		for i := itemsStartIdx; i < itemsEndIdx; i++ {
			result.Results = append(result.Results, response.Results[i])
		}
		result.Baton = response.Baton
		result.BaseUrl = response.BaseUrl

		if isEOF {
			break
		}
		sqlBatch, isEOF = chunker.Next()
	}
	return result, nil
}
```

Reply:
The chunker you pasted will return only one chunk with all statements in it, no? It also is not skipping over empty statements. I get the idea though. Let me try to implement it.

Reply:
Oops, yes - I forgot condition for `len` of a chunk